sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/ThroughputEstimator.java

```diff
@@ -55,25 +46,18 @@ public ThroughputEstimator() {
    * @param timeOfRecords the committed timestamp of the records
    * @param bytes the total bytes of the records
    */
+  @SuppressWarnings("nullness") // queue is never null, nor the peeked element
   public void update(Timestamp timeOfRecords, long bytes) {
     synchronized (queue) {
-      BigDecimal bytesNum = BigDecimal.valueOf(bytes);
-      if (startTimeOfCurrentWindow.equals(Timestamp.MIN_VALUE)) {
-        bytesInCurrentWindow = bytesNum;
-        startTimeOfCurrentWindow = timeOfRecords;
-        return;
-      }
-
-      if (timeOfRecords.getSeconds() < startTimeOfCurrentWindow.getSeconds() + 1) {
-        bytesInCurrentWindow = bytesInCurrentWindow.add(bytesNum);
+      if (queue.isEmpty() || timeOfRecords.getSeconds() > queue.peek().getLeft().getSeconds()) {
```

Comment:
`queue.peek()` will return the recent added element in the queue, right?

Reply:
Yes

Reply:
No, you are right, I should be using something else (maybe a stack).

Reply:
I think `queue` the right data structure. I just want to make sure those functions do exactly what we want.

Reply:
I think `queue.peak()` should be replaced by `queue.peakLast()`.

Reply:
`Deque` supports both operations for the front and the end of the queue. Updated the interface used, but the implementation class did not change (still using `ArrayDeque`).